SafeFoodie_Fresh/test/Text_Field.dart

```diff
@@ -0,0 +1,301 @@
+                child: Form(
+                  key: formKey,
+                  child: TextFormField(
+                    key: inputKey,
+                    controller: controller,
+                    // initialValue is 'Plover'
+                  ),
+                ),
+              ),
+            ),
+          ),
+        ),
+      );
+    }
+
+    await tester.pumpWidget(builder());
+    await tester.showKeyboard(find.byType(TextFormField));
+    final EditableTextState editableText =
+        tester.state(find.byType(EditableText));
+
+    // overwrite initial value.
+    controller.text = 'Xyzzy';
+    await tester.idle();
+    expect(editableText.widget.controller.text, equals('Xyzzy'));
+    expect(inputKey.currentState!.value, equals('Xyzzy'));
+    expect(controller.text, equals('Xyzzy'));
+
+    // verify value resets to initialValue on reset.
+    formKey.currentState!.reset();
+    await tester.idle();
+    expect(inputKey.currentState!.value, equals('Plover'));
+    expect(editableText.widget.controller.text, equals('Plover'));
+    expect(controller.text, equals('Plover'));
+  });
+
+  testWidgets('No crash when a TextFormField is removed from the tree',
+      (WidgetTester tester) async {
+    final GlobalKey<FormState> formKey = GlobalKey<FormState>();
+    String? fieldValue;
+
+    Widget builder(bool remove) {
+      return MaterialApp(
+        home: MediaQuery(
+          data: const MediaQueryData(),
+          child: Directionality(
+            textDirection: TextDirection.ltr,
+            child: Center(
+              child: Material(
+                child: Form(
+                  key: formKey,
+                  child: remove
+                      ? Container()
+                      : TextFormField(
+                          autofocus: true,
+                          onSaved: (String? value) {
+                            fieldValue = value;
+                          },
+                          validator: (String? value) {
+                            return (value == null || value.isEmpty)
+                                ? null
+                                : 'yes';
+                          },
+                        ),
+                ),
+              ),
+            ),
+          ),
+        ),
+      );
+    }
+
+    await tester.pumpWidget(builder(false));
+
+    expect(fieldValue, isNull);
+    expect(formKey.currentState!.validate(), isTrue);
+
+    await tester.enterText(find.byType(TextFormField), 'Test');
+    await tester.pumpWidget(builder(false));
+
+    // Form wasn't saved yet.
+    expect(fieldValue, null);
+    expect(formKey.currentState!.validate(), isFalse);
+
+    formKey.currentState!.save();
+
+    // Now fieldValue is saved.
+    expect(fieldValue, 'Test');
+    expect(formKey.currentState!.validate(), isFalse);
+
+    // Now remove the field with an error.
+    await tester.pumpWidget(builder(true));
+
+    // Reset the form. Should not crash.
```

Comment:
is there a function that catches it if it crashed during the reset?

Reply:
Nope there is no function to catch it if it crashes we should add one in the future.